src/backend/distributed/planner/recursive_planning.c

```diff
@@ -1337,6 +1349,255 @@ NodeContainsSubqueryReferencingOuterQuery(Node *node)
+		OpExpr *opExpr = (OpExpr *) node;
+		Var *leftColumn = LeftColumnOrNULL(opExpr);
+		Var *rightColumn = RightColumnOrNULL(opExpr);
+		UpdateVarNosInQualForSubquery((Node *) leftColumn);
+		UpdateVarNosInQualForSubquery((Node *) rightColumn);
+	}
+	else if (IsA(node, BoolExpr))
+	{
+		BoolExpr *boolExpr = (BoolExpr *) node;
+		List *argumentList = boolExpr->args;
+
+		Node *arg = NULL;
+		foreach_ptr(arg, argumentList)
+		{
+			UpdateVarNosInQualForSubquery(arg);
+		}
+	}
+}
+
+
+/*
+ * ContainsTableToBeConvertedToSubquery checks if the given range table list contains
+ * any table that should be converted to a subquery, which otherwise is not plannable.
+ */
+bool
+ContainsTableToBeConvertedToSubquery(List *rangeTableList, Oid resultRelationId)
+{
+	if (AllDataLocallyAccessible(rangeTableList))
+	{
+		return false;
+	}
+	if (ContainsLocalTableDistributedTableJoin(rangeTableList))
+	{
+		return true;
+	}
+	if (ContainsLocalTableSubqueryJoin(rangeTableList, resultRelationId))
+	{
+		return true;
+	}
+	return false;
+}
+
+
+/*
+ * AllDataLocallyAccessible return true if all data for the relations in the
+ * rangeTableList is locally accessible.
+ */
+static bool
+AllDataLocallyAccessible(List *rangeTableList)
+{
+	RangeTblEntry *rangeTableEntry = NULL;
+	foreach_ptr(rangeTableEntry, rangeTableList)
+	{
+		if (rangeTableEntry->rtekind == RTE_SUBQUERY)
+		{
+			/* TODO:: check if it has distributed table */
+			return false;
+		}
+		if (!IsRecursivelyPlannableRelation(rangeTableEntry))
+		{
+			continue;
+		}
+
+
+		Oid relationId = rangeTableEntry->relid;
+
+		if (!IsCitusTable(relationId))
+		{
+			/* local tables are locally accessible */
+			continue;
+		}
+
+		List *shardIntervalList = LoadShardIntervalList(relationId);
+		if (list_length(shardIntervalList) != 1)
+		{
+			/* we currently only consider single placement tables */
+			return false;
+		}
+
+		ShardInterval *shardInterval = linitial(shardIntervalList);
+		uint64 shardId = shardInterval->shardId;
+		ShardPlacement *localShardPlacement =
+			ShardPlacementOnGroup(shardId, GetLocalGroupId());
+		if (localShardPlacement == NULL)
+		{
+			/* the table doesn't have a placement on this node */
+			return false;
+		}
+	}
+
+	return true;
+}
+
+
+/*
+ * IsRecursivelyPlannableRelation returns true if the given range table entry
+ * is a relation type that can be converted to a subquery.
+ */
+bool
+IsRecursivelyPlannableRelation(RangeTblEntry *rangeTableEntry)
+{
+	if (rangeTableEntry->rtekind != RTE_RELATION)
+	{
+		return false;
+	}
+	return rangeTableEntry->relkind == RELKIND_PARTITIONED_TABLE ||
+		   rangeTableEntry->relkind == RELKIND_RELATION ||
+		   rangeTableEntry->relkind == RELKIND_MATVIEW ||
+		   rangeTableEntry->relkind == RELKIND_FOREIGN_TABLE;
+}
+
+
+/*
+ * ContainsLocalTableDistributedTableJoin returns true if the input range table list
+ * contains a direct join between local and distributed tables.
+ */
+bool
+ContainsLocalTableDistributedTableJoin(List *rangeTableList)
```

Comment:
Here if there is only reference and citus local tables, we might not need to convert it to a subquery, so we should also handle that I guess